### ISSUE TITLE
Missing NodeId when traversing the tree and doing some additional checks 

### ISSUE THREAD

**Issue by cecton** [iwburns/id-tree#63]
I'm trying to traverse the tree using one of the method (`PreOrderTraversal`) but I'm missing the NodeId in order to implement my code correctly.

## Code

```rust
let root_id = self.tree.root_node_id().unwrap();
for node in self.tree.traverse_pre_order(&root_id).unwrap() {
    let entity = node.data();
    let state = self.store.get_state();

    if !entity.is_active(state) {
        continue;
    }
    if let Some(parent) = node.parent() {
        if !self.tree.get(parent).unwrap().data().is_active(state) {
            continue
        }
        if self.tree.ancestors(parent).unwrap().any(|x| !x.data().is_active(state)) {
            continue
        }
    }

    entity.render(&mut self.app.canvas, state, &mut self.resources);
}
```

I kept the not relevant variables in this example in order to understand why I'm doing this.

This code works but is overly complicated.

## Explanation

* I have a tree of boxed entities (`Box<Entity>`).
* `Entity` is a trait that has 2 methods: `render()` and `is_active()`.
* An `Entity` is rendered (`render()`) only if its active (`is_active()` returns `true`) and all its ancestors are active.
* (An `Entity` must not render **if** it's active but the parent isn't or any ancestor isn't.)

## My problem

I don't have the `NodeId` from the iterator so I'm forget to get the parent (there I have a `NodeId`) check the parent and then call `tree.ancestors` to iterate over the rest of the ancestors. This code would be much simpler if I had a mean to get the `NodeId`'s of a `Node` or if I could iterate over `NodeId` instead of `Node`.

What would you suggest me to do in this scenario?

Another idea I had was to implement my own traversal method, so I copy-pasted [the code](https://docs.rs/id_tree/1.2.0/src/id_tree/iterators.rs.html#141-144) of `PreOrderTraversal` in my own code in order to modify it but it failed because `tree.nodes` (and `tree.get_unsafe`) are private.

If I can make my own traversal method I would probably be able to avoid pushing the children in the iterator if the parent isn't active.

**Comment by iwburns** [comment]
First off, thanks so much for submitting an issue!

If you'd like to be able to iterate over the `NodeId`s directly that's certainly possible.  If you decide to go this route would you mind submitting a PR?  I can assist wherever necessary, but I've been busy lately and don't have the time to actually write the code right now.

However, I think implementing your own traversal would be more ideal for this scenario (as you said, ideally you don't want to deal with the children of inactive nodes in the iterator at all).  If I were going to do this I'd start with the code from `PreOrderTraversal` and then do a few things to it:
1. Either of the following:
  a. remove the pre-allocation in `PreOrderTraversal::new`.
  b. submit a PR with a new function: `Tree.capacity` which will give you what you want in your custom traversal.
2. And also either of the following:
  a. ignore the fact that `Tree.get` is slightly less than optimal in terms of performance and just use it instead of `Tree.get_unsafe`
  b. submit a PR with a new function: `pub unsafe fn get_unchecked()` which is fully public (and marked as unsafe).  This function would behave just like `Tree.get_unsafe` (minus the internal unsafe block).

Option A in both of those is the easier but less performant option, while both option Bs require library changes (which is unfortunate, but I think these changes are not too extreme).

Please let me know if you have any questions about the above; or, if you decide to submit a PR I'd be happy to help wherever I can with that as well.

**Comment by cecton** [comment]
I was hoping that I missed something useful in the lib that would solve all my problems to be honest 😀 

Yes actually when writing it I realized that making my own traversal would make more sense. So, adapting id_tree to be able to do so is probably a more generic approach that people can reuse and benefit (while iterating over NodeId is not really an invitation to do a good code in this particular case).

Now to make my own iterator I would probably want to customize it if I want the vector to be allocated at the size of the tree or not. I can imagine that in some scenarios you would want the opposite behavior. So adding a `Tree.capacity()` method would be beneficial. (1.b. -> I will make you a PR for that)

Using `get_unsafe` is completely unnecessary in my opinion. Plus since the result of the `next` function is an `Option<T>`, why not directly using this for the result (it's wrapped in an option for free after all), I don't know if Rust will allow me though. (2.a. -> I will not add `Tree.get_unchecked`)

Now I have never made a PR for Rust and I actually started Rust recently so... I hope I will produce something decent 😁
